tests/integration_tests/noctx/testcases/st20p_user_pacing_tests.cpp

```diff
@@ -0,0 +1,119 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright(c) 2025 Intel Corporation
+ */
+
+#include <memory>
+
+#include "noctx.hpp"
+
+TEST_F(NoCtxTest, st20p_default_timestamps) {
+  initSt20pDefaultContext();
+
+  auto txBundle = createSt20pHandlerBundle(
+      /*createTx=*/true, /*createRx=*/false,
+      [](St20pHandler* handler) { return new FrameTestStrategy(handler); });
+  auto rxBundle = createSt20pHandlerBundle(
+      /*createTx=*/false, /*createRx=*/true,
+      [](St20pHandler* handler) { return new St20pDefaultTimestamp(handler); });
+  auto* frameTestStrategy = static_cast<St20pDefaultTimestamp*>(rxBundle.strategy);
+
+  ASSERT_TRUE(startRxThenTx(rxBundle, txBundle));
+
+  TestPtpSourceSinceEpoch(nullptr);
+  mtl_start(ctx->handle);
+  sleepUntilFailure();
+
+  stopTxThenRx(txBundle, rxBundle);
+
+  ASSERT_GT(frameTestStrategy->idx_rx, 0u)
+      << "st20p_user_pacing did not receive any frames";
+}
+
+TEST_F(NoCtxTest, st20p_user_pacing) {
+  initSt20pDefaultContext();
+
+  auto pacingState = std::make_shared<StrategySharedState>();
+
+  auto handlerTxBundle = createSt20pHandlerBundle(
+      /*createTx=*/true, /*createRx=*/false,
+      [](St20pHandler* handler) { return new St20pUserTimestamp(handler); },
+      [](St20pHandler* handler) {
+        handler->sessionsOpsTx.flags |= ST20P_TX_FLAG_USER_PACING;
+      },
+      pacingState);
+
+  auto handlerRxBundle = createSt20pHandlerBundle(
+      /*createTx=*/false, /*createRx=*/true,
+      [](St20pHandler* handler) { return new St20pUserTimestamp(handler); }, nullptr,
+      pacingState);
+
+  auto* frameTestStrategyTx = static_cast<St20pUserTimestamp*>(handlerTxBundle.strategy);
+  auto* frameTestStrategyRx = static_cast<St20pUserTimestamp*>(handlerRxBundle.strategy);
+
+  ASSERT_TRUE(startRxThenTx(handlerRxBundle, handlerTxBundle));
+
+  TestPtpSourceSinceEpoch(nullptr);
+  mtl_start(ctx->handle);
+
+  ASSERT_EQ(frameTestStrategyTx->getPacingParameters(), 0);
+  EXPECT_GT(frameTestStrategyTx->pacing_tr_offset_ns, 0.0);
+  EXPECT_GT(frameTestStrategyTx->pacing_trs_ns, 0.0);
+  EXPECT_GT(frameTestStrategyTx->pacing_vrx_pkts, 0u);
+
+  sleepUntilFailure();
+
+  stopTxThenRx(handlerTxBundle, handlerRxBundle);
+
+  ASSERT_GT(frameTestStrategyTx->idx_tx, 0u)
+      << "st20p_user_pacing did not transmit any frames";
+  ASSERT_GT(frameTestStrategyRx->idx_rx, 0u)
+      << "st20p_user_pacing did not receive any frames";
+  ASSERT_EQ(frameTestStrategyTx->idx_tx, frameTestStrategyRx->idx_rx)
+      << "TX/RX frame count mismatch";
+}
```

Comment:
i gennerally think this is overcomplicated, 

why not create just the St20Handler as one element,
add the funcitons that will do the cleanup in the default rx, 
i would generally just modify the st20 default function to do somthing along the line of 

if (stop_flag){
do once
  initalizaiton period (to make sure tx is intialized)

test_function()
}

until (deinitalization_period) {
test_function()
}


I would not use the shared state at all, as the frameTestStrategy would suffice, then it would all be in one element 

If done this way this whole class could be removed. As the frameStrategy or it's child that would work on pacing for you would serve that function without the shared object

Reply:
addressed. just one handler is used now.